### ISSUE TITLE
knxd incompatibility with vhci-hcd and new kernels on Raspberry Pi

### ISSUE THREAD

**Issue by MMartinezV** [knxd/knxd#528]
Hi,

I've been using eibd and later knxd for many years with a USB JUNG interface. Time ago I dockerized my setup to easy the upgrades and yesterday, trying to add a new funcionality to my setup I found a problem.

Now I want to read information about a modbus RTU device (a Huawei Inverter). To do it, because of the distance between my main Raspberry Pi which is connected to the KNX bus and the modbus connection, i decided to use USBIP (https://derushadigital.com/other%20projects/2019/02/19/RPi-USBIP-ZWave.html). 

To follow the setup, I need to install usbip on the Raspberry PI with Debian/Buster which runs KNXD. There is no problem with that. Then I need to load the vhci-hcd module and that causes a problem. After loading it knxd crashes and is not able to write on the bus. I can restart the docker container but after few seconds, on the first write attempt to the KNX bus it crashes.

If I unload the kernel module vhci-hcd, my setup works again, so I think it's clear that there is a problem with that.

I thought that maybe updating the system to Rasbian/Bullseye could solve the problem but no, it get's worse. With Raspbian/Bullseye my setup fails always, it does not depend on vhci-hcd module, and KNXD crashes as soon as it tries to do a write on the KNX bus.

More information. 

* System version: Rasbian/buster with kernel 5.10.17-v7l+
* KNXD version: 0.14.29
* Docker-compose knxd setup
<pre><code>
knxd:
    image: welteki/knxd:0.14.29
    restart: always
    command: knxd /config.ini
    volumes:
      - "./knxd/config.ini:/config.ini"
      - "./knxd/70-knxd.rules:/etc/udev/rules.d/knxd/70-knxd.rules"
    devices:
      - /dev/bus/usb:/dev/bus/usb
    network_mode: "host"
</code></pre>

knxd/config.ini
<code><pre>
[tcp]
port=6720
server=knxd_tcp
systemd-ignore=false

;[ip]
;driver=ip

[usb-con]
driver=usb
;send-timeout=6000

[debug-main]
error-level=0x9
trace-mask=0xffc

[debug-server]
name=mcast:knxd

[main]
addr=1.1.128
client-addrs=1.1.129:8
connections=server,tcp,usb-con
debug=debug-main
systemd=systemd

[server]
debug=debug-server
discover=true
router=router
server=ets_router
tunnel=tunnel
</pre></code>

knxd/70-knxd.rules
<pre><code>
SUBSYSTEM=="usb", ATTRS{idVendor}=="135e", ATTRS{idProduct}=="0023", SYMLINK+="ttyKNX1", OWNER="knxd", MODE="0664", ACTION=="add"
</code></pre>

I've read diferent issues related to knxd crashes when writing on the bus 
* https://community.openhab.org/t/knxd-crashes-after-sending-to-usb/103924
* https://github.com/knxd/knxd/issues/517
* https://github.com/knxd/knxd/issues/415

Probably they are all quite similar and I hope this information can help to find the problem.

Regards,

  Manuel Martínez

**Comment by d-buchmann** [comment]
I can confirm this issue after flashing the latest openHABian (based on bullseye) to my Raspberry Pi 3.

My KNX hardware is a Pigator module, which ran fine for the last 4 years or so, but I had to do a reinstall because of a failed storage device. I followed the install guide for this HW from the knxd wiki and am pretty sure it's set up correctly.
I can open `/dev/ttyKNX1` with minicom and get output when a device sends something (but I couldn't test sending).

`$ uname -a: Linux openhabian 5.10.92-v7+ #1514 SMP Mon Jan 17 17:36:39 GMT 2022 armv7l GNU/Linux`
`$ knxd -V: v0.14.46`

I will try to do some debugging, but I can't promise anything. Also, my C++ experience is rather limited.

**Comment by d-buchmann** [comment]
Just tested with `v0.14.54-1` compiled from the debian branch, same result.
BTW, for me the "result" expresses as:
<pre><code>
$ knxd -t1023 -f9 -DRTS -e 1.1.128 -E 1.1.129:1 -b tpuarts:/dev/ttyKNX1
Layer 3 [ 4:server/Server      0.000] registerLink: 4:server
Layer 3 [ 4:server/Server       0.001] Start: cfg:server
Layer 5 [ 4:server/Server       0.001] down => >up
Layer 8 [ 4:server/Server       0.001] Open
Layer 0 [ 4:server/Server       0.001] Open
Layer 0 [ 4:server/Server       0.001] Opened
Layer 8 [17:router/driver       0.001] OpenD
Layer 8 [17:router/driver       0.002] OpenedD
N00000127: [18:router.pace_] The 'pace' filter without a queue acts globally.
Layer 3 [16:router/ConnC        0.002] registerLink: 16:router_16
Layer 3 [16:router/ConnC        0.002] Start: cfg:router
Layer 5 [16:router/ConnC        0.002] down => >up
Layer 5 [16:router/ConnC        0.002] Starting
Layer 5 [16:router/ConnC        0.002] >up => up
Layer 4 [16:router/ConnC        0.002] link state changed: up
Layer 5 [16:router/ConnC        0.002] Started
Layer 4 [16:router/ConnC        0.002] link state changed: up
Layer 8 [ 4:server/Server       0.002] Opened
Layer 5 [ 4:server/Server       0.002] >up => up
Layer 4 [ 4:server/Server       0.002] link state changed: up
Layer 5 [ 4:server/Server       0.002] Started
Layer 4 [ 4:server/Server       0.003] link state changed: up
Layer 0 [ 4:server/Server       0.751] Recv(016): 06 10 02 07 00 10 01 00 08 01 C0 A8 B2 22 AF 28
Layer 2 [ 4:server/Server       0.752] Unknown connection 1
Layer 1 [ 4:server/Server       0.752] Send(002): 01 21
Layer 0 [ 4:server/Server       0.752] Send(008): 06 10 02 08 00 08 01 21
F00000105: [11:A.tpuarts] Link down, terminating
Layer 4 [16:router/ConnC        2.493] R Stopping
Layer 5 [16:router/ConnC        2.493] up => >down
Layer 5 [16:router/ConnC        2.493] L Stopping
Layer 5 [16:router/ConnC        2.493] >down => down
Layer 4 [16:router/ConnC        2.493] link state changed: down
Layer 4 [16:router/ConnC        2.493] link state changed: down
Layer 4 [ 4:server/Server       2.493] R Stopping
Layer 5 [ 4:server/Server       2.493] up => >down
Layer 5 [16:router/ConnC        2.493] L Stopping
Layer 3 [16:router/ConnC        2.494] unregisterLink: router_16
Layer 0 [ 4:server/Server       2.494] Close D
Layer 5 [ 4:server/Server       2.494] >down => down
Layer 4 [ 4:server/Server       2.494] link state changed: down
Layer 4 [ 4:server/Server       2.494] link state changed: down
Layer 8 [17:router/driver       2.494] CloseD
Layer 4 [ 4:server/Server       2.494] is down
Layer 8 [ 4:server/Server       2.494] Close E
</code></pre>

**Comment by d-buchmann** [comment]
Meanwhile, I've tested another Raspi with a new TPUART interface on a vanilla Raspbian OS "legacy" (aka buster). Why I'm posting this here is because my first suspicion was the OS release, because that was basically the only thing that had changed for me.

Now, I even tested the Raspberry UART with a loopback cable and it won't send a single char...same as the first Raspi with the recent OS, and I can't imagine how I possibly could have managed to fry both devices' UART TX pins...
Anyways, my case seems to be closed here (hardware issue), and sorry @MMartinezV for hijacking this issue for no reason after all.

**Comment by MMartinezV** [comment]
Bad news. Meanwhile I've switched to calimero server and everything is working fine.

Thanks!

**Comment by MMartinezV** [comment]
Hi again,

After many months using Calimero server with OpenHAB, today I've tried again to use the latest dockerized version of knxd (welteki/knxd:0.14.54), and it seems to work fine.

I've done some changes last year. I switched to 32 bit Raspbian Bullseye and recently upgraded OH 3.3 to OH 3.4.

I've not done tests, but as I still have the old SD card with Raspbian64 with OH 3.3 and a previous version of welteki/knxd, I'll try update the docker image of knxd and see if it works in that setup.
